File: game304/deck.py
"""Deck management: creation, shuffling, cutting, and dealing.

The 304 pack consists of 32 cards (ranks 7–A in four suits). Shuffling
deliberately preserves partial card order from the previous game to
encourage higher bids — this is achieved via overhand-style shuffling
(1–4 partial passes). A full Fisher-Yates shuffle is used only after
three consecutive reshuffles (a house rule safety valve).
"""

from __future__ import annotations

import random

from game304.card import Card
from game304.seating import deal_order
from game304.types import Rank, Seat, Suit


def create_pack() -> list[Card]:
    """Create a fresh 32-card pack in standard order.

    Cards are ordered by suit (Clubs, Diamonds, Hearts, Spades),
    then by rank within each suit (J, 9, A, 10, K, Q, 8, 7).

    Returns:
        A list of 32 ``Card`` instances.
    """
    return [Card(rank, suit) for suit in Suit for rank in Rank]


class Deck:
    """A mutable deck of cards with shuffle, cut, and deal operations.

    All randomness is routed through an injectable ``random.Random``
    instance, making the deck fully deterministic when seeded. This
    is critical for reproducible tests.

    Args:
        cards: Initial cards in the deck. If ``None``, a fresh
            32-card pack is created.
        rng: Random number generator. Defaults to an unseeded
            ``random.Random()`` if not provided.

    Examples:
        >>> deck = Deck(rng=random.Random(42))
        >>> deck.minimal_shuffle()
        >>> hands = deck.deal(Seat.NORTH, 4)
        >>> len(hands[Seat.WEST])  # first to receive
        4
    """

    def __init__(
        self,
        cards: list[Card] | None = None,
        rng: random.Random | None = None,
    ) -> None:
        self._cards: list[Card] = cards if cards is not None else create_pack()
        self._rng: random.Random = rng if rng is not None else random.Random()

    @property
    def cards(self) -> list[Card]:
        """The current ordered list of cards in the deck."""
        return self._cards

    def __len__(self) -> int:
        return len(self._cards)

    # ------------------------------------------------------------------
    # Shuffling
    # ------------------------------------------------------------------

    def overhand_shuffle(self) -> None:
        """Perform a single overhand-style partial shuffle.

        Splits the deck into random chunks of 2–6 cards from the top
        and prepends each chunk to a result pile. This mimics a real
        overhand shuffle, preserving some card adjacency.

        Per the rules: "Minimal shuffling is key — this preserves the
        order of cards from the previous game to encourage high-value
        betting."
        """
        result: list[Card] = []
        remaining = list(self._cards)
        while remaining:
            chunk_size = min(
                2 + self._rng.randint(0, 4),
                len(remaining),
            )
            chunk = remaining[:chunk_size]
            remaining = remaining[chunk_size:]
            # Prepend chunk to result (place on top)
            result = chunk + result
        self._cards = result

    def minimal_shuffle(self) -> None:
        """Perform 1–4 overhand shuffles (minimal shuffling per rules).

        The number of passes is chosen randomly. This preserves
        partial card order from the previous game.
        """
        num_passes = 1 + self._rng.randint(0, 3)
        for _ in range(num_passes):
            self.overhand_shuffle()

    def full_shuffle(self) -> None:
        """Perform a Fisher-Yates shuffle (complete randomisation).

        Used after 3 consecutive reshuffles to reset the pack
        (house rule). This destroys all card-order information.
        """
        self._rng.shuffle(self._cards)

    # ------------------------------------------------------------------
    # Cutting
    # ------------------------------------------------------------------

    def cut(self) -> None:
        """Cut the deck at a random point.

        The cut point is chosen uniformly between index 1 and
        ``len - 2`` (inclusive), ensuring at least one card remains
        on each side.

        Per the rules: "The player to the left of the dealer may cut
        the pack once, or tap the top of the deck to decline the cut."
        """
        if len(self._cards) < 3:
            return
        cut_point = 1 + self._rng.randint(0, len(self._cards) - 3)
        self._cards = self._cards[cut_point:] + self._cards[:cut_point]

    # ------------------------------------------------------------------
    # Dealing
    # ------------------------------------------------------------------

    def deal(self, dealer: Seat, num_cards: int) -> dict[Seat, list[Card]]:
        """Deal cards from the top of the deck to each player.

        Cards are dealt one at a time in anticlockwise order, starting
        with the player to the dealer's right. Each player receives
        ``num_cards`` cards. Dealt cards are removed from the deck.

        Per the rules: "The dealer then deals the top 4 cards to the
        player on their right, then 4 to the next player
        (anticlockwise), and so on."

        Note: The JS implementation deals all ``num_cards`` to each
        player before moving to the next. The rules say "the top 4
        cards to the player on their right, then 4 to the next" which
        means each player gets a batch, not interleaved. We follow the
        JS implementation's batched dealing approach.

        Args:
            dealer: The seat of the dealer.
            num_cards: Number of cards to deal to each player.

        Returns:
            A dict mapping each ``Seat`` to their dealt cards.

        Raises:
            ValueError: If there aren't enough cards in the deck.
        """
        total_needed = num_cards * 4
        if len(self._cards) < total_needed:
            raise ValueError(
                f"Not enough cards to deal: need {total_needed}, "
                f"have {len(self._cards)}"
            )

        order = deal_order(dealer)
        hands: dict[Seat, list[Card]] = {seat: [] for seat in order}

        for _ in range(num_cards):
            for seat in order:
                hands[seat].append(self._cards.pop(0))

        return hands

File: frontend/src/304dle/engine/deck.ts
// Deck management: shuffling, cutting, and dealing.
// Mirrors game304/deck.py.
//
// Note on RNG: Python uses Mersenne Twister; this port uses
// xorshift32 (already used by 304dle's dealing.ts). The two RNGs
// produce different sequences from the same numeric seed — so the
// TS Deck is *internally deterministic* given a TS RNG, but does
// NOT bit-match Python's deck for the same numeric seed. Engine
// parity tests inject pre-shuffled card orders from Python fixtures
// and bypass the in-engine shuffle, so this divergence does not
// affect parity verification.

import type { CardId } from './card';
import { PACK } from './card';
import type { Seat } from './seating';
import { dealOrder } from './seating';

// ---------------------------------------------------------------------------
// Random — small wrapper that mirrors the slice of Python's random.Random
// the engine uses (randint inclusive, shuffle).
// ---------------------------------------------------------------------------

export interface Random {
  // Inclusive on both ends, mirroring Python's random.Random.randint.
  randint(low: number, high: number): number;
  // In-place Fisher-Yates shuffle.
  shuffle<T>(arr: T[]): void;
}

// xorshift32 — fast, deterministic, browser-safe.
export const makeRandom = (seed: number): Random => {
  let s = seed | 0;
  if (s === 0) s = 1;
  const next01 = (): number => {
    s ^= s << 13;
    s ^= s >>> 17;
    s ^= s << 5;
    s = s | 0;
    return (s >>> 0) / 0x100000000;
  };
  return {
    randint(low, high) {
      const span = high - low + 1;
      return low + Math.floor(next01() * span);
    },
    shuffle<T>(arr: T[]) {
      for (let i = arr.length - 1; i > 0; i--) {
        const j = Math.floor(next01() * (i + 1));
        const tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
      }
    },
  };
};

// ---------------------------------------------------------------------------
// Pack
// ---------------------------------------------------------------------------

export const createPack = (): CardId[] => [...PACK];

// ---------------------------------------------------------------------------
// Deck
// ---------------------------------------------------------------------------

export class Deck {
  private _cards: CardId[];
  private _rng: Random;

  constructor(cards?: ReadonlyArray<CardId>, rng?: Random) {
    this._cards = cards ? [...cards] : createPack();
    this._rng = rng ?? makeRandom(Date.now() | 0);
  }

  get cards(): ReadonlyArray<CardId> {
    return this._cards;
  }

  get remaining(): number {
    return this._cards.length;
  }

  // Single overhand-style partial shuffle. Splits the deck into
  // chunks of 2–6 from the top and prepends each chunk to the
  // result pile (mirroring a real overhand shuffle, preserving
  // some adjacency).
  overhandShuffle(): void {
    const result: CardId[] = [];
    let remaining = this._cards;
    while (remaining.length > 0) {
      const chunkSize = Math.min(
        2 + this._rng.randint(0, 4),
        remaining.length,
      );
      const chunk = remaining.slice(0, chunkSize);
      remaining = remaining.slice(chunkSize);
      // Prepend chunk to result (place on top).
      result.splice(0, 0, ...chunk);
    }
    this._cards = result;
  }

  // 1–4 overhand shuffles.
  minimalShuffle(): void {
    const numPasses = 1 + this._rng.randint(0, 3);
    for (let i = 0; i < numPasses; i++) this.overhandShuffle();
  }

  // Full Fisher-Yates randomisation (used after 3 consecutive reshuffles).
  fullShuffle(): void {
    this._rng.shuffle(this._cards);
  }

  // Cut at a random point (between 1 and len-2 inclusive).
  cut(): void {
    if (this._cards.length < 3) return;
    const cutPoint = 1 + this._rng.randint(0, this._cards.length - 3);
    this._cards = [
      ...this._cards.slice(cutPoint),
      ...this._cards.slice(0, cutPoint),
    ];
  }

  // Pop a single card from the top of the deck.
  pop(): CardId | undefined {
    return this._cards.shift();
  }

  // Pop n cards from the top of the deck.
  popN(n: number): CardId[] {
    if (n > this._cards.length) {
      throw new Error(
        `Not enough cards in deck: requested ${n}, have ${this._cards.length}`,
      );
    }
    return this._cards.splice(0, n);
  }

  // Deal `numCards` to each seat in dealing order (anticlockwise from
  // dealer's right). Each seat receives a contiguous batch before the
  // next seat. Cards are removed from the deck.
  deal(dealer: Seat, numCards: number): Record<Seat, CardId[]> {
    const totalNeeded = numCards * 4;
    if (this._cards.length < totalNeeded) {
      throw new Error(
        `Not enough cards to deal: need ${totalNeeded}, have ${this._cards.length}`,
      );
    }

    const order = dealOrder(dealer);
    const hands: Record<Seat, CardId[]> = {
      north: [],
      west: [],
      south: [],
      east: [],
    };
    for (const seat of order) {
      hands[seat] = this._cards.splice(0, numCards);
    }
    return hands;
  }

  // Inject a specific card order. Used by tests / parity fixtures to
  // bypass shuffling and exercise downstream logic with a known deck.
  setOrder(cards: ReadonlyArray<CardId>): void {
    this._cards = [...cards];
  }
}
